CHI_TECH/ChiMath/Quadratures/quadrature_gausslegendre.cc

```diff
@@ -12,64 +20,123 @@ void chi_math::QuadratureGaussLegendre::
   {
     printf("Initializing Gauss-Legendre Quadrature with %d q-points\n",N);
   }
-  //============================================= InitializeAlphaElements init guess and weights
-  double dx = 2.0/2000;
-  double xgold = -1.0;
-  double fold = Legendre(N,xgold);
-  //printf("Fold=%f\n",fold);
-  double xgnew = 0.0;
-  double fnew = 0.0;
-  for (int i=0;i<2000;i++)
+
+  //========================= Compute the roots
+  FindRoots(N, abscissae, maxiters, tol);
+
+  //========================= Compute the weights
+  weights.resize(N,1.0);
+  for (size_t k=0; k < abscissae.size(); k++)
   {
-    xgnew = xgold + dx;
-    fnew = Legendre(N,xgnew);
-    //printf("xgnew=%f Fnew=%f  Fold=%f\n",xgnew,fnew,fold);
-    if ((fnew*fold)<0.0)
-    {
-      abscissae.push_back(xgnew);
-      weights.push_back(1.0);
-    }
+    weights[k] =
+      2.0 * (1.0 - abscissae[k] * abscissae[k]) /
+      ( (N + 1) * (N + 1) *
+        Legendre(N+1, abscissae[k]) * Legendre(N+1, abscissae[k]) );
 
-    xgold = xgnew;
-    fold = fnew;
+    if (verbose)
+      chi_log.Log(LOG_0)
+        << "root[" << k << "]=" << abscissae[k]
+        << ", weight=" << weights[k];
+  }//for abscissae
+}
+
+//###################################################################
+/** Finds the roots of the Legendre polynomial.
+ *
+ * The algorithm is that depicted in:
+ *
+ * [1] Barrera-Figueroa, et al., "Multiple root finder algorithm for Legendre
+ *     and Chebyshev polynomials via Newton's method", Annales Mathematicae et
+ *     Informaticae, 33 (2006) pp. 3-13.
+ *
+ * \param N Is the order of the polynomial.
+ * \param roots Is a reference to the roots.
+ * \param max_iters Maximum newton iterations to perform for each root.
+ *        Default: 1000.
+ * \param tol Tolerance at which the newton iteration will be terminated.
+ *        Default: 1.0e-12.
+ *
+ * \author Jan*/
+void chi_math::QuadratureGaussLegendre::FindRoots(
+  int N, std::vector<double> &roots, size_t max_iters, double tol)
```

Comment:
IDK. Can't remember? Why not? Maybe cause it should be >=0. What would you like it to be? int ..? done

Reply:
I like that more too. Let's do that.

Reply:
> IDK. Can't remember? Why not? Maybe cause it should be >=0. What would you like it to be? int ..? done

This is something we should probably discuss and include in our style guide.  I tend to try to avoid unsigned integers if possible (unless we're using them in the context of addresses or sizes). I think it creates more problems than it solves.  This sort of sums it up for me: https://wesmckinney.com/blog/avoid-unsigned-integers/.